fix: 32bit-ARMv7 crashes on android

NoelStephensUnity opened the pull request:
This PR resolves the issue with ARMv7 builds crashing on the Android.

fix: #2646

## Changelog

- Fixed: Issue where ARMv7 Android builds would crash when trying to validate the batch header.

## Testing and Documentation

- Includes integration tests (?)
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
temp fix

Temporarily commenting out the batch header hash check to have a user validate that this fixes the issue with Android Armv7 builds.

NoelStephensUnity committed:
fix

This resolves the issue with XXHash causing ARMv7 builds to crash when validating the batched message header and payload.
Keep messages sent word aligned.
Use the 32bit version of XXHash.

NoelStephensUnity committed:
fix

Prevent padding on the NetworkBatchHeader

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Messaging/CustomMessageManager.cs:
```diff
@@ -199,14 +199,18 @@ internal void InvokeNamedMessage(ulong hash, ulong sender, FastBufferReader read
-
-            m_MessageHandlerNameLookup32[hash32] = name;
-            m_MessageHandlerNameLookup64[hash64] = name;
+            if (m_NetworkManager.NetworkConfig.RpcHashSize == HashSize.VarIntFourBytes)
```
I don't think this should be configurable, personally... There's not really any significant advantage to using a 64-bit hash over a 32-bit hash here. Slightly higher chance with 32 bits that two values will have the same hash, but for what we use it for (a debugging tool to detect corruption) and given the other methods we also use of detecting corruption, I think we should just always use a 32-bit hash.

Making it configurable creates room for user error (both in terms of using the wrong one for the target platform, and in terms of trying to do it differently on different platforms and resulting in platforms being unable to talk to each other) that we'd have to provide support on.

NoelStephensUnity replied:
The issue is that CustomMessageManager had half of the script using that configuration and the other half ignored it... 
So, either we remove it completely or we keep it consistent throughout the script.  
I am ok with either.

ShadauxCat replied:
Oh. I misread this change. I thought it was the corruption check hash in NetworkMessageManager. Carry on.

ShadauxCat committed:
Added logic on the receive end to account for the extra padding, added some extra safety logic on both read and write sides.

ShadauxCat committed:
Fixed tests

NoelStephensUnity committed:
Merge branch 'develop' into fix/32bit-custom-named-message-crash-android

NoelStephensUnity committed:
style

Minor style adjustment

NoelStephensUnity committed:
style

fixing whitespace issue with the other test changes.

ShadauxCat committed:
Move alignment to only happen at the batch level

NoelStephensUnity committed:
fix

This resolves the tools bytes measured tests.

NoelStephensUnity committed:
fix

This resolves the tools bytes measured tests.

ShadauxCat committed:
Test fixes

ShadauxCat committed:
Merge remote-tracking branch 'origin/armv7_kitty_test' into armv7_kitty_test

NoelStephensUnity committed:
update

switching back to 64 bit hash values for batch header validation.

NoelStephensUnity committed:
Merge branch 'armv7_kitty_test' into fix/32bit-custom-named-message-crash-android

ShadauxCat commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/TransportBytesMetricsTests.cs:
```diff
@@ -71,7 +74,11 @@ public IEnumerator TrackTotalNumberOfBytesReceived()
             Assert.True(observer.Found);
-            Assert.AreEqual(FastBufferWriter.GetWriteSize(messageName) + k_MessageOverhead, observer.Value);
+
+            var measuredSize = (long)(Math.Ceiling(observer.Value * 1.0f / 8.0f) * 8.0f);
```
I don't think this value should be adjusted. It should already be word-aligned.

NoelStephensUnity committed:
test revert

reverting the switch to 32 bit hash values.

NoelStephensUnity committed:
update revert

reverting the conditional registration of either 32bit or 64 bit hash values in either table to just register in both tables.

NoelStephensUnity committed:
style

 reverting the additional CR/LF

NoelStephensUnity committed:
style

reverting the code to be exactly as it was (no need to make organizational changes)

NoelStephensUnity committed:
test

removing the alignment calculation from the observed byte count.
Removing additional CR/LFs.

NoelStephensUnity committed:
style

removing extra LF/CRs.

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/TransportBytesMetricsTests.cs:
```diff
@@ -41,7 +41,8 @@ public IEnumerator TrackTotalNumberOfBytesSent()
 
             Assert.True(observer.Found);
-            Assert.AreEqual(FastBufferWriter.GetWriteSize(messageName) + k_MessageOverhead, observer.Value);
+            var expectedSize = (long)(Math.Ceiling((FastBufferWriter.GetWriteSize(messageName) + k_MessageOverhead) * (1.0f / 8.0f)) * 8.0f);
```
any reason why we're not doing the 1/8 precalculation here too?

simon-lemay-unity commented:
Will let you decide if you think it's worth the hit to readability, but you can calculate aligned lengths this way:
```c
alignedLength = (length + 7) & ~7;
```
It avoids dealing with floating points and only uses fast integer instructions.

ShadauxCat commented:
> Will let you decide if you think it's worth the hit to readability, but you can calculate aligned lengths this way:
> 
> ```c
> alignedLength = (length + 7) & ~7;
> ```
> 
> It avoids dealing with floating points and only uses fast integer instructions.

Oh good point. +1 to this @NoelStephensUnity

SamuelBellomo commented on com.unity.netcode.gameobjects/Runtime/Messaging/NetworkMessageManager.cs:
```diff
@@ -829,11 +830,17 @@ internal unsafe void ProcessSendQueues()
-                    queueItem.BatchHeader.BatchSize = queueItem.Writer.Length;
+
+                    var alignedLength = (int)(Math.Ceiling(queueItem.Writer.Length * k_WordAlignBatchCalc) * 8);
+                    queueItem.Writer.TryBeginWrite(alignedLength);
```
(ramping up on this, let me know if I'm completely misunderstanding)
Do I understand correctly that this basically adds a few bits for each batched message? Do we know how much overhead this adds when you have a lot of batched messages? (wondering if it'd be worth it to ifdef this just for Android ARMv7)

ShadauxCat replied:
It adds a few bytes (0-7 bytes) per batch of messages, not per message.

Unfortunately we can't do that #if because every platform has to speak the same protocol. Can't have one platform add extra bytes to a message that the other platform isn't accounting for on the receiving end. Plus, it's actually the receiving end that needs this, so Windows has to add these bytes if sending to Android... and we don't have any information on whether or not the remote end is an android device.

NoelStephensUnity replied:
Just for context:
The user scenario was an ARMv7 build running on an Android device connecting to a headless 64bit linux server build (I tested with 64bit windows, but the idea is the same).

SamuelBellomo replied:
ah true right.
nvm then resolving

NoelStephensUnity committed:
update

Applying Simon's suggestion.

NoelStephensUnity committed:
update

This update assures we won't go above the MTU size due to word alignment.

NoelStephensUnity committed:
style

removing VS 2022 auto-assigned namespace...<grr>
